src/Integration/VsSessionHost.cs

```diff
@@ -296,6 +295,65 @@ public object GetService(Type type)
         }
         #endregion
 
+        #region IServiceContainer
+        void IServiceContainer.AddService(Type serviceType, object serviceInstance)
+        {
+            ((IServiceContainer)this).AddService(serviceType, serviceInstance, false);
+        }
+
+        void IServiceContainer.AddService(Type serviceType, object serviceInstance, bool promote)
+        {
+            if (serviceInstance == null)
+            {
+                throw new ArgumentNullException(nameof(serviceInstance));
+            }
+
+            ((IServiceContainer)this).AddService(serviceType, (c,t)=> serviceInstance, promote);
+        }
+
+        void IServiceContainer.AddService(Type serviceType, ServiceCreatorCallback callback)
+        {
+            ((IServiceContainer)this).AddService(serviceType, callback, false);
+        }
+
+        void IServiceContainer.AddService(Type serviceType, ServiceCreatorCallback callback, bool promote)
+        {
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            Debug.Assert(!promote, "Not expecting to use promote these services");
+            Debug.Assert(!this.localServices.ContainsKey(serviceType), string.Format(CultureInfo.CurrentCulture, $"Service {serviceType.FullName} is already registered"));
+            Debug.Assert(typeof(ILocalService).IsAssignableFrom(serviceType), "Unexpected service registration");
+
+            this.localServices.Add(serviceType, new Lazy<ILocalService>(() => (ILocalService)callback(this, serviceType)));
+        }
+
+        void IServiceContainer.RemoveService(Type serviceType)
+        {
+            ((IServiceContainer)this).RemoveService(serviceType, false);
+        }
+
+        void IServiceContainer.RemoveService(Type serviceType, bool promote)
+        {
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            Debug.Assert(!promote, "Not expecting to use promote these services");
+            Debug.Assert(this.localServices.ContainsKey(serviceType), string.Format(CultureInfo.CurrentCulture, $"Service {serviceType.FullName} is not registered"));
+
+            this.localServices.Remove(serviceType);
+        }
+        #endregion
+
```

Comment:
Could you use the framework ServiceContainer class instead of providing an implementation?
If the only reason to implement the interface is so that a service can be replaced by a test then this could be accomplished by either exposing the _serviceContainer_ as an internal, or adding a specific internal method e.g. 
`internal ReplaceInternalServiceForTesting<T>(T newService)`